### ISSUE TITLE
[REVIEW] Error when creating bash_utils.sh from Docker

### ISSUE THREAD

**Issue by santisoler** [underworldcode/stripy#19]
After installing `stripy` using Docker, I'm following the installation instructions for creating and installing the helper scripts for bash.
So, with the `underworldcode/stripy:latest` container already pulled, I run:
```
docker run --rm underworldcode/stripy:latest bash_utils > bash_utils.sh
```
But, the `bash_utils.sh` contains only the following line:
```
/usr/bin/xvfb-run: line 184: bash_utils: not found
```

Do you have any idea what could be failing?

**System specifications:**
- OS: Manjaro Linux XFCE 64bits
- Docker version 18.09.5-ce, build e8ff056dbc

**Comment by lmoresi** [comment]
Hmmm ... we may be missing something in the tool chain that gets loaded up when building the docker image. I’ll need to check that again.

Prof Louis Moresi

louis.moresi@unimelb.edu.au<mailto:louis.moresi@unimelb.edu.au>

(w) +61 3 8344 1217

(m) +61 4 0333 1413

(us) +1 505 349 4425

www.moresi.info<http://www.moresi.info/>

www.facebook.com/underworldcode<http://www.facebook.com/underworldcode>

@LouisMoresi<https://twitter.com/LouisMoresi>


On 25 Apr 2019, 12:26 -0700, Santiago Soler <notifications@github.com>, wrote:

After installing stripy using Docker, I'm following the installation instructions for creating and installing the helper scripts for bash.
So, with the underworldcode/stripy:latest container already pulled, I run:

docker run --rm underworldcode/stripy:latest bash_utils > bash_utils.sh


But, the bash_utils.sh contains only the following line:

/usr/bin/xvfb-run: line 184: bash_utils: not found


Do you have any idea what could be failing?

System specifications:

  *   OS: Manjaro Linux XFCE 64bits
  *   Docker version 18.09.5-ce, build e8ff056dbc

**Comment by lmoresi** [comment]
Ok a it’s a missing package from the dockerfile that must have got lost when we switched to a different base image. I’ll fix it up. How does this work with the submission / release ?

Prof Louis Moresi

louis.moresi@unimelb.edu.au<mailto:louis.moresi@unimelb.edu.au>

(w) +61 3 8344 1217

(m) +61 4 0333 1413

(us) +1 505 349 4425

www.moresi.info<http://www.moresi.info/>

www.facebook.com/underworldcode<http://www.facebook.com/underworldcode>

@LouisMoresi<https://twitter.com/LouisMoresi>


On 25 Apr 2019, 12:26 -0700, Santiago Soler <notifications@github.com>, wrote:

After installing stripy using Docker, I'm following the installation instructions for creating and installing the helper scripts for bash.
So, with the underworldcode/stripy:latest container already pulled, I run:

docker run --rm underworldcode/stripy:latest bash_utils > bash_utils.sh


But, the bash_utils.sh contains only the following line:

/usr/bin/xvfb-run: line 184: bash_utils: not found


Do you have any idea what could be failing?

System specifications:

  *   OS: Manjaro Linux XFCE 64bits
  *   Docker version 18.09.5-ce, build e8ff056dbc

**Comment by santisoler** [comment]
> Ok a it’s a missing package from the dockerfile that must have got lost when we switched to a different base image. I’ll fix it up. How does this work with the submission / release ?

So it wasn't a big issue, great! :+1:
If you have to make a change on the repo in order to fix the Docker, you could start a Pull Request with the fixing and then merge it into the `paper` branch.
This way we can keep track of which changes fixes which issues (if you write `Fixes #...` on the PR description, the issue gets closed when the PR is merged - see how I referenced #17 on #18).
If you don't know how to do it, you could just commit the changes into the `paper` branch, no big deal.

I'm thinking we could apply future changes to the current `paper` branch and after acceptance you could make a new release. I'll ask the editors for guidance on this subject.

**Comment by lmoresi** [comment]
So, I have updated the docker image. It's not always obvious how to track versioning through this process. The changes are reflected in the dockerfile that we (happen to) keep in our repository. I pushed this to the master branch and merged into the paper branch. To test the docker help you will need to re-pull the docker image. As we haven't talked much about the versioning for this, I haven't changed the docker tag or the README.md in the repo ... happy to take advice on the best way to reflect these kinds of changes in the manuscript versioning

**Comment by santisoler** [comment]
> So, I have updated the docker image. It's not always obvious how to track versioning through this process. The changes are reflected in the dockerfile that we (happen to) keep in our repository. I pushed this to the master branch and merged into the paper branch. To test the docker help you will need to re-pull the docker image. 

That's great @lmoresi, that should do the trick.
Perhaps is cleaner and simpler to make this changes directly on the `paper` branch, so you don't clutter it with the latest changes in `master`.
Nevertheless, I think this issues are related with how your git workflow has been defined (see #16 for my review about it). There's no big problems regarding it, but I think it will eventually give you some headaches.

I've just pulled the new Docker image.
The `bash_utils.sh` is now correctly created, although I'm getting the following error when I run `stripy-docker-browse` after sourcing the bash script.
```
Navigate to http://localhost:8899 to view the examples, ^\ when done
"docker run" requires at least 1 argument.
See 'docker run --help'.

Usage:  docker run [OPTIONS] IMAGE [COMMAND] [ARG...]

Run a command in a new container
```
Seems that the Docker name is missing on the command lines inside `bash_utils.sh`. Am I right?

**Comment by santisoler** [comment]
> As we haven't talked much about the versioning for this, I haven't changed the docker tag or the README.md in the repo ... happy to take advice on the best way to reflect these kinds of changes in the manuscript versioning

I'm not familiar with tagging Docker containers, but seems that following the same version number than the one assigned to GitHub Releases is the best choice.
Nevertheless, for cases like this where you change the Docker container but don't make a GitHub Release,  you could use the hash of the commit for versioning your Docker. This allows you to keep track of which commit has created that Docker container. [This article](https://container-solutions.com/tagging-docker-images-the-right-way/) recommends this approach and gives more insight on how to do it.

I'm still thinking we should not care too much about versioning on the review process. We should just apply the changes on the `paper` branch (or on the `master` branch if you agree with #16) by merging Pull Requests or just by pushing commits to it.
After acceptance you could create a new release and then tag it as the *final* reviewed version.
I'm still waiting for guidance from the editors on this subject, but don't worry too much about it for now.

**Comment by lmoresi** [comment]
I'm not sure - I wonder if it is my inadequate understanding of bash functions. I tidied up a little bit and re-enabled the command-line killing option which had stopped working. If you can `docker pull` again, would be great. 

I have this as the bash function that gets defined after the templating process spits out the bash_utils.sh. My version of bash is` GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin18)
Copyright (C) 2007 Free Software Foundation, Inc`

```
# Open the default version of the docker to browse the examples
stripy-docker-browse(){
  PORT=${1:-8899} ;
  echo "Navigate to http://localhost:${PORT} to view the examples, ^\ when done" ;
  docker run -it --rm -p ${PORT}:8888 -v ${PWD}:/home/jovyan/external underworldcode/stripy:0.7 ;
}
```

**Comment by santisoler** [comment]
I'm not used to working with Docker, so my lack of experience didn't let me see an obvious mistake.
That's why I limited myself to run the instructions on the `README.md`.

I've just realized that the `docker pull underworldcode/stripy:latest` didn't pull the `underworldcode/stripy:0.7` version. I thought the `latest` version was just a pointer to the newest tag, like GitHub's `latest` version works. My bad, sorry for the inconvenience.

Now `bash_utils.sh` matches the one on `paper` branch: [`paper:Docker/scripts/bash_utils.sh`](https://github.com/underworldcode/stripy/blob/master/Docker/scripts/bash_utils.sh).

I think this issue is solved, so feel free to close it when you're ready.
Thanks for helping me to solve this.
